developer/src/kmc-kmn/src/ng-compiler/scan-recognizer.ts

```diff
@@ -0,0 +1,243 @@
+/*
+ * Keyman is copyright (C) SIL Global. MIT License.
+ *
+ * Created by Dr Mark C. Sinclair on 2025-11-05
+ *
+ * KMC KMN Next Generation Lexer
+ *
+ * ScanRecognizer for the Lexer
+ */
+
+import { TokenType } from "./token-type.js";
+
+/**
+ * A ScanRecognizer identifies an individual Token as part of the Next Generation Lexer.
+ */
+export class ScanRecognizer {
+  /**
+   * Construct a ScanRecognizer
+   */
+  public constructor(
+    /** the token type to return if matched */
+    public readonly tokenType: TokenType,
+    /** the regex to identify the token */
+    public readonly regExp: RegExp,
+    /**  whether to emit the token or not? */
+    public readonly emit: boolean
+  ) {
+    this.regExp = new RegExp(regExp); // does not preserve lastIndex
+  }
+
+  public toString(): string {
+    return `[${this.tokenType},${this.regExp},${this.emit}]`;
+  }
+}
+
+// the ordering of ScanRecognizers is important, with more specific regexs appearing first
+
+export const KMN_SCAN_RECOGNIZERS = [
+  // https://help.keyman.com/developer/language/reference/baselayout
```

Comment:
Now wondering if these comments should be string properties of the scan recognizer, which could give us easy lookups for error messages in the future? (and allow us to even automatically verify links are not 404 too, at least in theory)